Implement `SquashDictInnerCheckAccessIndex` hint

TAdev0 opened the pull request:
Resolves #298 

This PR implements `SquashDictInnerCheckAccessIndex` hint, which updates the access index during the dictionary squashing process

TAdev0 commented:
@rodrigo-pino  ready for review

har777 commented on pkg/hintrunner/zero/zerohint_dictionaries.go:
```diff
@@ -89,6 +89,78 @@ func createSquashDictInnerAssertLenKeysHinter() (hinter.Hinter, error) {
+// during the dictionary squashing process
+//
+// `newSquashDictInnerCheckAccessIndexHint` takes 1 operander as argument
+//   - `loopTemps` variable is a struct containing a `index_delta_minus1` field
```
can you also say `as the first field`
or maybe address it as a comment below in the code somewhere so the reason for writing to offset `0` of struct is clear

TAdev0 replied:
@har777  just addressed your comment, tell me if it is better like this

har777 commented on pkg/hintrunner/zero/zerohint_dictionaries.go:
```diff
@@ -154,6 +154,81 @@ func createSquashDictInnerAssertLenKeysHinter() (hinter.Hinter, error) {
+
+			currentAccessIndex, ok := currentAccessIndex_.(fp.Element)
+			if !ok {
+				return fmt.Errorf("casting currentAccessIndices_ into an array of felts failed")
```
you mean into a felt not an array of felts

TAdev0 replied:
true thanks !

har777 commented on pkg/hintrunner/zero/zerohint_dictionaries.go:
```diff
@@ -154,6 +154,81 @@ func createSquashDictInnerAssertLenKeysHinter() (hinter.Hinter, error) {
+
+			// We use 0 as offset for `WriteToNthStructField` function as we write
+			// to the first field of the `loop_temps` struct
+			return vm.Memory.WriteToNthStructField(loopTempsAddr, resultMem, int16(0))
```
If the offset is `0` you can just use `WriteToAddress(loopTempsAddr, resultMem)` also

TAdev0 replied:
true. It seems more straightforward, will do that. thanks!

har777 commented on pkg/hintrunner/zero/zerohint_dictionaries.go:
```diff
@@ -154,6 +154,81 @@ func createSquashDictInnerAssertLenKeysHinter() (hinter.Hinter, error) {
+				return fmt.Errorf("casting currentAccessIndices_ into an array of felts failed")
+			}
+
+			newAccessIndex, err := utils.Pop(&currentAccessIndices)
+			if err != nil {
+				return err
+			}
```
I believe you need to update `current_access_indices` in the scope with the new popped array?

har777 replied:
As `.pop()` updates the `current_access_indices` object.

TAdev0 replied:
oh damn i forgot that i thought i dit it , and indeed its missing. thanks!

har777 commented on pkg/hintrunner/zero/zerohint_dictionaries_test.go:
```diff
@@ -96,6 +96,49 @@ func TestZeroHintDictionaries(t *testing.T) {
+				},
+				check: func(t *testing.T, ctx *hintTestContext) {
+					varValueEquals("loop_temps.index_delta_minus1", feltUint64(0))(t, ctx)
+					varValueInScopeEquals("current_access_index", *feltUint64(2))(t, ctx)
```
can you also add a check for `current_access_indices` in scope after the pop

TAdev0 replied:
yes right!

TAdev0 commented:
@har777  just addressed all your comments! thanks for the feedback. Indeed the hint was wrongly implemented, it didnt update the array of felts in scope after popping an element.
I added a test for `current_access_indices` array in scope, now it should be good :)

MaksymMalicki commented on pkg/hintrunner/zero/zerohint_dictionaries.go:
```diff
@@ -396,6 +396,85 @@ func createSquashDictInnerAssertLenKeysHinter() (hinter.Hinter, error) {
+//
+// `newSquashDictInnerCheckAccessIndexHint` writes to the first field `index_delta_minus1`
+// of the `loop_temps` struct the result of `new_access_index - current_access_index - 1`
+// and assigns `new_access_index` to `current_access_index` in the scope
```
could you add to this annotation an information about how does the loop_temps struct looks like?

TAdev0 replied:
sure

MaksymMalicki commented on pkg/hintrunner/zero/zerohint_dictionaries.go:
```diff
@@ -396,6 +396,85 @@ func createSquashDictInnerAssertLenKeysHinter() (hinter.Hinter, error) {
+
+			currentAccessIndices, ok := currentAccessIndices_.([]fp.Element)
+			if !ok {
+				return fmt.Errorf("casting currentAccessIndices_ into a felt failed")
```
felt array 😛

TAdev0 replied:
i inversed with another one yeah thanks!

MaksymMalicki commented on pkg/hintrunner/zero/zerohint_dictionaries_test.go:
```diff
@@ -245,6 +245,49 @@ func TestZeroHintDictionaries(t *testing.T) {
+					varValueEquals("loop_temps.index_delta_minus1", feltUint64(23))(t, ctx)
+					allVarValueInScopeEquals(map[string]any{"current_access_index": *feltUint64(43), "current_access_indices": []fp.Element{*feltUint64(97), *feltUint64(76), *feltUint64(54), *feltUint64(51), *feltUint64(44)}})(t, ctx)
+				}},
+		},
```
is it a good ida to test a case with empty `current_access_indices` ?

TAdev0 replied:
just added one more test for that :)

TAdev0 commented:
@MaksymMalicki addressed your comments
